### PR TITLE
WIN32: hint noreturn functions

### PR DESCRIPTION
**Pull request by jaltman**
apply __declspec(noreturn) compiler hints to functions that
do not return.

Change-Id: I3c6c4703c7235d1df3e21dccad5272ea4cddfd36

### CODE REVIEW

**Comment by nicowilliams** [comment]
LGTM!

**Comment by nicowilliams** [comment]
Ah, but something is wrong.  From the Travis builds:

```
  CC       cms.lo
In file included from ../../../lib/hx509/cert.c:34:
In file included from ../../../lib/hx509/hx_locl.h:98:
../../../lib/hx509/hx509-private.h:13:12: error: expected function body after function declarator
        noreturn) _hx509_abort(const char */*fmt*/,
                  ^
../../../lib/hx509/cert.c:353:2: warning: implicit declaration of function '_hx509_abort' is invalid in C99 [-Wimplicit-function-declaration]
        _hx509_abort("cert refcount <= 0 on free");
        ^
In file included from ../../../lib/hx509/ca.c:34:
In file included from ../../../lib/hx509/hx_locl.h:98:
../../../lib/hx509/hx509-private.h:13:12: error: ../../../lib/hx509/cert.c:3409:1: error: conflicting types for '_hx509_abort'
_hx509_abort(const char *fmt, ...)
^
```

File: lib/hx509/cert.c [hunk 1]
```diff
@@ -3403,6 +3403,9 @@ hx509_cert_binary(hx509_context context, hx509_cert c, heim_octet_string *os)
 #define __attribute__(X)
 
 void
+#ifdef _WIN32
+_declspec(noreturn)
```
Comment (nicowilliams) [inline lib/hx509/cert.c]: Missing underscore?

Comment (nicowilliams) [reply lib/hx509/cert.c]: And whitespace issue?

Comment (nicowilliams) [reply lib/hx509/cert.c]: Also, why are we `#define`ing a no-op `__attribute__()` macro above, at 3403??  Shouldn't we be doing that only in the `_WIN32` case??

**Comment by jaltman** [comment]
@nicowilliams This series fixes the exports from libhx509 on windows and adds the noreturn hinting.   It does not remove the unconditional definition of `__attribute__(X)` from lib/hx509/cert.c because doing so as in ef038c6 results in appveyor build failures.    The macro is not defined in the source file at that point.   Perhaps you can track that down independently after this series is merged.

**Comment by nicowilliams** [comment]
Let's look at warnings on appveyor.

First, a minor annoyance:

```
--------- Entering lib roken:
	cl -Zi -Od -DDEBUG /Zi -c -DCRTAPI1=_cdecl -DCRTAPI2=_cdecl -nologo -GS -D_AMD64_=1 -DWIN64 -D_WIN64  -DWIN32 -D_WIN32 -W4 -D_WINNT -D_WIN32_WINNT=0x0501 -DNTDDI_VERSION=0x05010000 -D_WIN32_IE=0x0600 -DWINVER=0x0501  -D_MT -D_DLL -MDd    -DHAVE_CONFIG_H -DHCRYPTO_DEF_PROVIDER=w32crypto  -IC:\projects\heimdal\lib\roken\..\..\out\dest_AMD64\inc -IC:\projects\heimdal\lib\roken\..\..\out\dest_AMD64\inc\krb5   -D_CRT_SECURE_NO_WARNINGS -wd4996 -wd4127 -wd4244 -wd4100 -wd4706 -wd4214 -wd4267 -wd4018 -wd4389 -wd4204 -wd4221 -wd4295 -wd4146 /MP .\base64-test.c  -FoC:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\ -FdC:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\ -DBUILD_ROKEN_LIB
base64-test.c
C:\projects\heimdal\out\dest_AMD64\inc\roken.h(61) : warning C4005: 'ROKEN_LIB_NORETURN_FUNCTION' : macro redefinition
        C:\projects\heimdal\out\dest_AMD64\inc\config.h(92) : see previous definition of 'ROKEN_LIB_NORETURN_FUNCTION'
	link /DEBUG /DEBUGTYPE:cv /DEBUG  /INCREMENTAL:NO /NOLOGO -subsystem:console,5.01 kernel32.lib  ws2_32.lib mswsock.lib advapi32.lib /machine:X64 -OUT:C:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\base64-test.exe C:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\base64-test.obj C:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\base64.obj C:\projects\heimdal\lib\roken\..\..\out\dest_AMD64\lib\libroken.lib
```

Next, let's check on uninitialized variable warnings...  We now see only four of those!  Yes!

```
.\test_context.c(761) : warning C4133: 'function' : incompatible types - from 'krb5_enctype *' to 'int32_t *'
c:\projects\heimdal\lib\gssapi\test_context.c(248) : warning C4701: potentially uninitialized local variable 'output_token' used
c:\projects\heimdal\lib\gssapi\test_context.c(263) : warning C4701: potentially uninitialized local variable 'actual_mech_client' used
c:\projects\heimdal\lib\gssapi\test_context.c(263) : warning C4701: potentially uninitialized local variable 'actual_mech_server' used
```
and
```
c:\projects\heimdal\lib\asn1\check-gen.c(1498) : warning C4701: potentially uninitialized local variable 'size' used
```

**Comment by jaltman** [comment]
> 
> 
> Let's look at warnings on appveyor.
> 
> First, a minor annoyance:
> 
> ```
> --------- Entering lib roken:
> 	cl -Zi -Od -DDEBUG /Zi -c -DCRTAPI1=_cdecl -DCRTAPI2=_cdecl -nologo -GS -D_AMD64_=1 -DWIN64 -D_WIN64  -DWIN32 -D_WIN32 -W4 -D_WINNT -D_WIN32_WINNT=0x0501 -DNTDDI_VERSION=0x05010000 -D_WIN32_IE=0x0600 -DWINVER=0x0501  -D_MT -D_DLL -MDd    -DHAVE_CONFIG_H -DHCRYPTO_DEF_PROVIDER=w32crypto  -IC:\projects\heimdal\lib\roken\..\..\out\dest_AMD64\inc -IC:\projects\heimdal\lib\roken\..\..\out\dest_AMD64\inc\krb5   -D_CRT_SECURE_NO_WARNINGS -wd4996 -wd4127 -wd4244 -wd4100 -wd4706 -wd4214 -wd4267 -wd4018 -wd4389 -wd4204 -wd4221 -wd4295 -wd4146 /MP .\base64-test.c  -FoC:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\ -FdC:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\ -DBUILD_ROKEN_LIB
> base64-test.c
> C:\projects\heimdal\out\dest_AMD64\inc\roken.h(61) : warning C4005: 'ROKEN_LIB_NORETURN_FUNCTION' : macro redefinition
>         C:\projects\heimdal\out\dest_AMD64\inc\config.h(92) : see previous definition of 'ROKEN_LIB_NORETURN_FUNCTION'
> 	link /DEBUG /DEBUGTYPE:cv /DEBUG  /INCREMENTAL:NO /NOLOGO -subsystem:console,5.01 kernel32.lib  ws2_32.lib mswsock.lib advapi32.lib /machine:X64 -OUT:C:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\base64-test.exe C:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\base64-test.obj C:\projects\heimdal\lib\roken\..\..\out\obj_AMD64\lib\roken\base64.obj C:\projects\heimdal\lib\roken\..\..\out\dest_AMD64\lib\libroken.lib
> ```
> 
> There's a bunch of those. It's probably an easy fix.

I'm working on that now.

**Comment by nicowilliams** [comment]
The uninitialized variable warnings in `lib/gssapi/test_context.c` are lame.  I dunno why VC thinks they might be used uninitialized -- it must not be fully-analyzing that `while` loop.

**Comment by nicowilliams** [comment]
The warning in `lib/asn1/check-gen.c` is the sort of warning one would think your patches would fix:

```
static int
check_TESTMechTypeList(void)
{
    TESTMechTypeList tl;
    unsigned oid1[] =  { 1, 2, 840, 48018, 1, 2, 2};
    unsigned oid2[] =  { 1, 2, 840, 113554, 1, 2, 2};
    unsigned oid3[] =   { 1, 3, 6, 1, 4, 1, 311, 2, 2, 30};
    unsigned oid4[] =   { 1, 3, 6, 1, 4, 1, 311, 2, 2, 10};
    TESTMechType array[] = {{ 7, oid1 },
                            { 7, oid2 },
                            { 10, oid3 },
                            { 10, oid4 }};
    size_t size, len;
    void *ptr;
    int ret;

    tl.len = 4;
    tl.val = array;

    ASN1_MALLOC_ENCODE(TESTMechTypeList, ptr, len, &tl, &size, ret);
    if (ret)
        errx(1, "TESTMechTypeList: %d", ret);
    if (len != size)
        abort();
    return 0;
}
```
where the `ASN1_MALLOC_ENCODE()` macro is:
```
#define ASN1_MALLOC_ENCODE(T, B, BL, S, L, R)                  \
  do {                                                         \
    (BL) = length_##T((S));                                    \
    (B) = malloc((BL));                                        \
    if((B) == NULL) {                                          \
      (R) = ENOMEM;                                            \
    } else {                                                   \
      (R) = encode_##T(((unsigned char*)(B)) + (BL) - 1, (BL), \
                       (S), (L));                              \
      if((R) != 0) {                                           \
        free((B));                                             \
        (B) = NULL;                                            \
      }                                                        \
    }                                                          \
  } while (0)
```
and `encode_TESTMechTypeList()` always either returns an error or initializes that `size_t` output parameter.  So the only way this variable is used uninitialized is if `abort()` returns.  At least MinGW's `<stdlib.h>` declares `abort()` as `__cdecl __declspec(noreturn)`, but I've not checked Windows'.

**Comment by jaltman** [comment]
> 
> 
> The uninitialized variable warnings in `lib/gssapi/test_context.c` are lame. I dunno why VC thinks they might be used uninitialized -- it must not be fully-analyzing that `while` loop.

Its probably due to 
`        if (client_done && server_done)
            break;`
within the loop.   The compiler cannot determine that `server_done` is only non-zero if the loop fully executes at least once.  Hence, the warning is generate for the use of `output_token.length` at line 248 and not line 240.

**Comment by nicowilliams** [comment]
Yeah, that's what I was thinking, that VC doesn't analyze that loop far  enough.

BTW, you still have the macro redefinition warning:
```
C:\projects\heimdal\out\dest_AMD64\inc\roken.h(68) : warning C4005: 'ROKEN_LIB_NORETURN_FUNCTION' : macro redefinition
        C:\projects\heimdal\out\dest_AMD64\inc\config.h(92) : see previous definition of 'ROKEN_LIB_NORETURN_FUNCTION'
```
I don't mind those warnings, provided it doesn't cause VC to not emit other warnings.  Every time I fix warnings VC seems to add new ones, so I suspect it gives up after so many, which would mean we do want to avoid spurious warnings.

**Comment by jaltman** [comment]
> BTW, you still have the macro redefinition warning:
> 
> ```
> C:\projects\heimdal\out\dest_AMD64\inc\roken.h(68) : warning C4005: 'ROKEN_LIB_NORETURN_FUNCTION' : macro redefinition
>         C:\projects\heimdal\out\dest_AMD64\inc\config.h(92) : see previous definition of 'ROKEN_LIB_NORETURN_FUNCTION'
> ```
> 
> I don't mind those warnings, provided it doesn't cause VC to not emit other warnings. Every time I fix warnings VC seems to add new ones, so I suspect it gives up after so many, which would mean we do want to avoid spurious warnings.

The warnings are potentially serious since they are caused by an incorrectly generated `roken.h` from `roken.h.in`.   I hope that 73915cc does the right thing.

**Comment by jaltman** [comment]
73915cc does the right thing.

**Comment by nicowilliams** [comment]
Thanks!

**Pull request merged**